keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -424,64 +424,60 @@ _This section must be completed when targeting beta graduation to a release._
 
 * **What specific metrics should inform a rollback?**
 
+  - job_sync_duration_seconds shows significantly more latency for Indexed Jobs.
+  - job_sync_total shows errors for Indexed Jobs only.
+  - job_finished_total shows that Indexed Jobs don't finish.
+
 * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
-  Describe manual testing that was done and the outcomes.
-  Longer term, we may want to require automated upgrade/rollback tests, but we
-  are missing a bunch of machinery and tooling and can't do that now.
+
+  Manual test:
+  
+  1. Deploy k8s 1.21 cluster
+  1. Upgrade to 1.22
+  1. Create Indexed Job with big number of completions and pods that run for ~10min.
+  1. Downgrade to 1.21. Verify that no new pods are created for the Indexed Job.
+  1. Upgrade to 1.22. Verify that new pods are created for Indexed Job.
 
 * **Is the rollout accompanied by any deprecations and/or removals of features, APIs, 
 fields of API types, flags, etc.?**
-  Even if applying deprecation policies, they may still surprise some users.
+
+  No
 
 ### Monitoring Requirements
 
 _This section must be completed when targeting beta graduation to a release._
 
 * **How can an operator determine if the feature is in use by workloads?**
-  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
-  checking if there are objects with field X set) may be a last resort. Avoid
-  logs or events for this purpose.
+
+  - job_sync_total has values for the label completion_mode=Indexed.
 
 * **What are the SLIs (Service Level Indicators) an operator can use to determine 
 the health of the service?**
-  - [ ] Metrics
-    - Metric name:
-    - [Optional] Aggregation method:
-    - Components exposing the metric:
-  - [ ] Other (treat as last resort)
-    - Details:
+
+  - [x] Metrics
+    - Metric name (all new): job_sync_duration_seconds, job_sync_total, job_finished_total
+    - Components exposing the metric: kube-controller-manager
 
 * **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
-  At a high level, this usually will be in the form of "high percentile of SLI
-  per day <= X". It's impossible to provide comprehensive guidance, but at the very
-  high level (needs more precise definitions) those may be things like:
-  - per-day percentage of API calls finishing with 5XX errors <= 1%
-  - 99% percentile over day of absolute value from (job creation time minus expected
-    job creation time) for cron job <= 10%
-  - 99,9% of /health requests per day finish with 200 code
+
+  - per-day percentage of job_sync_total with label result=error <= 1%
+  - 99% percentile over day for job_sync_duration_seconds is:
+    - <= 5s for single pod Jobs 
+    - <= 1m for parallel jobs with a few tens of pods.
```

Comment:
Do you have any metrics already? Few tens of pods sync shouldn't be longer than 10s.

Reply:
This is based off the pod creation SLOs, which are at 1s per pod https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos

In general, these number are very conservative, as we don't know in which kind of machine the control plane will be running.

Reply:
From @wojtek-t 

> What exactly "job_sync_duration_seconds" is reporting?

> Is it reporting the duration of a single "sync in a controller"? Or processing a job?
If the former, it should be fairly fast right? [I'm assuming we're creating pods themselves asynchronously?]

It is reporting a single sync in a controller. We create/delete pods in parallel, but we wait for all of them to succeed. What would be the recommendation? Can we simply assume that all of them would finish within 1s?

Reply:
> It is reporting a single sync in a controller. We create/delete pods in parallel, but we wait for all of them to succeed.

Thanks for explanation.
We can't really assume they will finish fast then, as we still have QPS limits on the client-side.

The problem I see here, is that if a single sync can take that long, then one very large job (or N very large jobs where N=number of workers of the controller) can actually starve everything else for a really long time. That's far from perfect.

What we should do is to ensure that in a single sync we create no more than N pods (say 500 - I think 500 is used by ReplicaSet controller). And that gives you much better predictability (you can say that a single sync always has to finish within M seconds).
That has a drawback that it adds some latency if you indeed would be able to create e.g. 10k pods at once, but I think the impact is relatively small and you solve the starvation problem.

Reply:
I'm fine with adding a limit to the number of created pods. Should we do it for deletions too (happens when users reduces parallelism or the job reached failure limits)?

Then for 500 pods, assuming 50 calls per second, we are talking about 10s for pods + 1s for Job update. Sounds reasonable? This will increase a bit with job tracking with finalizers, but we can discuss that in its beta graduation.

Reply:
Yes - we should do for deletions too.

O(10s) SLO sounds waaay better than 1m or 1h.

> This will increase a bit with job tracking with finalizers, but we can discuss that in its beta graduation.

It adds one additional call per 500 calls, right? So I would say it's negligible.

Reply:
Right, big jobs are barely affected. And indexed jobs can actually skip some calls :)

In any case, I'll add SLOs for parallelism=1, 50 and 500

Reply:
> Right, big jobs are barely affected. And indexed jobs can actually skip some calls :)

Smaller jobs are not affected at all, because they will never not create a 500-item batch.
So only large jobs may be affected.

Reply:
> In any case, I'll add SLOs for parallelism=1, 50 and 500

I would really like to have one SLO (without any steps).

Reply:
> Smaller jobs are not affected at all

we still have to remove finalizers from the pods and job (but we are going off topic for the purpose of this PR).

> I would really like to have one SLO (without any steps).

will do, assuming 500 operations limit.

Reply:
updated